embassy-rp: enable access to low-level PIO registers and values

ericseppanen opened the pull request:
Existing code doesn't really support using PIO with complex DMA configurations (e.g. chaining dma channels to automatically loop buffers).

I found that there were a few things I wanted from embassy-rp to make these designs easier:
1. Access to the DREQ values for PIO FIFOs, to set up DMA pacing
2. Access to the FIFO hardware address, to load into the DMA read/write address register
3. Access to the PIO registers for debugging

I have been dropping down to `rp-pac` to do these things, but that fits poorly with the rest of my (embassy) code, and breaks down when using generic parameters for the pio and sm instances.

Feedback welcome! Feel free to bikeshed names, docs, etc.

ericseppanen committed:
embassy-rp: allow access to PIO registers

Register access is useful when debugging PIO issues, such as stalled
FIFOs or state machines waiting for irqs.

ericseppanen committed:
embassy-rp: make StateMachine{Tx/Rx} dreq fn public

This value is required to configure DMA to/from the state machine FIFOs,
so granting access to the value is useful if a program wants to do
things embassy-rp doesn't support yet (such as chaining DMA channels).

ericseppanen committed:
embassy-rp: allow retrieving the PIO FIFO address

These values are needed to manually configure DMA transfers in or out of
a PIO state machine FIFO.

ericseppanen commented:
> Ideally, the high level interface (of both DMA and PIO) would be expanded to cover your usecase.

Thanks for the feedback!

PIO+DMA can be combined in so many ways, that unless we have a bunch of applications to look at, it seems pretty hard to guess at an interface that supports a wide variety of use cases.

And since any high-level interface has to meet semver requirements, it can't evolve very quickly if it needs to wait for a new breaking release every time a change needs to be made. So I imagine it would be easier (both for downstream projects and for embassy-rp semver-compatibility) to allow people to do what they want with the PIO registers and then look for common patterns in the projects that result. I'm open to being proved wrong! I'm just not sure how to gather more information from the current state of things.

> The pac version is not bound to the semver of embassy-rp, so I think we cannot put registers in the public interface of embassy-rp.

That's good to know. I'm confused though by the existence of `embassy_rp::dma::Channel::regs()`... doesn't that do the same thing? (it returns `rp_pac::dma::Channel`.)

ericseppanen committed:
add dma::Channel::new_no_interrupt

Some PIO programs use DMA but don't need a CPU interrupt. Add a
constructor that doesn't require an interrupt parameter.

Dirbaio commented:
Yes, please gate anything that exposes PAC types on `unstable-pac`. LGTM otherwise. 
Mark as ready to review when done.

ericseppanen commented:
Thanks for the nudge; this PR is now rebased on latest main branch.

A few notes for reviewers, due to things that have changed upstream since my first attempt:

1. I added a DMA constructor that skips interrupt setup, because I am using DMA in a way that does not require interrupts to the cpu-- it's feeding data to PIO instances in a loop forever. I'm not certain how to use the new `dma::Channel` `Mode` parameter in this configuration. It would be a lot of work to add a new `Mode` just to specialize this niche use case, but I acknowledge it could be confusing if someone used the `new_no_interrupt` constructor and then expected interrupt-based blocking functions to work.

2. In my original PR I added `StateMachineTx::fifo_address()` and `StateMachineRx::fifo_address()`, and in the meantime someone also added `StateMachine::rx_fifo_ptr()` and `StateMachine::rx_fifo_ptr()`. I find it convenient to have it at the lower level (plus it tidies some internal code), but if the redundancy bothers reviewers this commit is not strictly necessary. Also, perhaps the names should be more similar?

3. Access to the `rp_pac::pio::Pio` register block is gated by the `unstable-pac` feature, as requested.

Dirbaio commented:
Thanks!

> It would be a lot of work to add a new Mode just to specialize this niche use case, but I acknowledge it could be confusing if someone used the new_no_interrupt constructor and then expected interrupt-based blocking functions to work.

Maybe `new_blocking` shouldn't enable the interrupt, AFAICT it doesn't provide any function that uses it. Anyway this is orthogonal to this PR so let's just merge.
